File: ProjectEuler/Problems/Problem016.cs
﻿using Microsoft.Extensions.Logging;

namespace ProjectEuler.Problems
{
    public class Problem016 : Problem<int>
    {
        public Problem016(ILogger<Problem016> logger) : base(logger) { }

        protected override int CalculateAnswer()
        {
            var product = twoTo100;

            for (var i = 2; i <= 10; ++i)
            {
                product = MathHelpers.MultiplyIntegerStrings(product, twoTo100);
            }

            Logger.LogDebug("{product}", product);

            var answer = 0;
            for (var i = 0; i < product.Length; ++i)
            {
                answer += product[i] - '0';
            }

            return answer;
        }

        //2^100
        private readonly string twoTo100 = "1267650600228229401496703205376";
    }
}

File: ProjectEulerPython/problems/problem_016.py
from problems.problem import Problem

# 2^100
TWO_TO_100 = 1267650600228229401496703205376

class Problem016(Problem):

  def calculate_answer(self) -> int:
    product = TWO_TO_100
    for i in range(2, 11):
      product *= TWO_TO_100
    self.print_detail("Product:")
    self.print_detail(product)

    answer = 0
    product_str = str(product)
    for i in range(0, len(product_str)):
      answer += int(product_str[i])

    return answer